File: Roadmap/03 - ESTRUCTURAS DE DATOS/python/D3rk1us.py
"""
 * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
 *   los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos.
 *   (o el número de dígitos que quieras)
 * - También se debe proponer una operación de finalización del programa.
 */
"""

# Listas

nombres= ['Maria', 'Nushi', 'Jose', 'Wei']
apellidos = [] #Lista vacía

matriz = [
    [1, 2, 4, 6, 9],
    [10, 20, 40, 80],
    [11, 22, 33, 44],
]

nombres.sort()  # Ordenar
print(nombres)

apellidos.append('Smith') # Insertar
print(apellidos)

# Tuplas

tupla1 = "Prueba", 34214, 3.2
tupla2 = tupla1, "Tres", "Cuatro", 14

# Conjuntos

frutas = {'Naranja', 'Manzana', 'Pera', 'Uva', 'Granada'}
frutas.remove('Uva')    # Borrar
print(frutas)


# Diccionarios

usuario = {'nombre': 'Marc', 'apellido': 'Spector', 'oficio': 'Agente de Inteligencia'}
usuario['nombre'] = 'Steven'  # Reemplazar
print(usuario)



# /////////////////////////////////////  DIFICULTAD EXTRA (opcional)  /////////////////////////////////////


agenda = [] # Lista para guardar los contactos.
exit = True # Variable para mantener o terminar el bucle While.

# Funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.

def agregar():

    nombre = input("Introduce el nombre del contacto: ")
    numero = input("Introduce el número del contacto: ")

    try:
        # Comprueba si es un número y si tiene menos de 11 dígitos.
        if isinstance(int(numero), int) and len(numero) <= 11:
        
            return agenda.append({'nombre': nombre, 'numero': numero})

    except:

        print("El número no debe tener letras y no debe tener más de 11 dígitos.")
        agregar()

def actualizar():

    opcion = input("Deseas actualizar el 'nombre' o 'número': ")

    if opcion == 'nombre':

        nombre_actual = input("Indica el nombre actual del contacto: ")
        cambio_nombre = input("Indica el nombre al que deseas cambiarlo: ")
         
        for contacto in agenda:

            if contacto['nombre'] == nombre_actual:
                contacto['nombre'] = cambio_nombre
                return contacto

    elif opcion == 'número':
        
        numero_actual = input("Indica el número actual del contacto: ")
        cambio_numero = input("Indica el número al que deseas cambiarlo: ")
        
        try:
            
            if isinstance(int(cambio_numero), int) and len(cambio_numero) <= 11:
                
                for contacto in agenda:

                    if contacto['numero'] == numero_actual:
                        contacto['numero'] = cambio_numero
                        return contacto

        except:

            print("El número no debe tener letras y no debe tener más de 11 dígitos.")
            actualizar()

    else:
        print("Escoger entre 'nombre' o 'número'.")
    
    return 'El contacto no ha sido creado.'

def eliminar():
    
    posicion = 0
    contacto_eliminar = input("Nombre del contacto que se va a eliminar: ")

    for contacto in agenda:

        if contacto['nombre'] == contacto_eliminar:

            break

        posicion += 1

    agenda.pop(posicion)

    return agenda

def buscar():

    nombre = input("Introduce el nombre del contacto que deseas buscar: ")

    for contacto in agenda:

        if contacto['nombre'] == nombre:
            return contacto


# Menú de la agenda y parte principal.

while exit:

    menu = input("¿Qué desea hacer? 'agregar', 'buscar', 'actualizar', 'eliminar', 'mostrar' un contacto o 'salir': ")

    if menu == 'agregar':

        agregar()

    elif menu == 'actualizar':

        print(actualizar())


    elif menu == 'eliminar':

        print(eliminar())

    elif menu == 'buscar':

        print(buscar())
    
    elif menu == 'mostrar':

        print(agenda)
    
    elif menu == 'salir':

        exit = False
    
    else:
        print("Debes escoger entre las opciones indicadas: 'añadir', 'buscar'...")

File: Roadmap/03 - ESTRUCTURAS DE DATOS/dart/D3rk1us.dart
import 'dart:collection';
import 'dart:io';

void main() {

/*
 * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
*/

    // Lista

    print('\nList: ');
    List<int> numeros = [2, 5, 6, 87, 12, 532];
    print(numeros);

    numeros.insert(0, 1); // Inserción
    print(numeros);

    numeros.remove(87); // Borrado
    print(numeros);

    numeros[3] = 601; // Actualización
    print(numeros);

    numeros.sort(); // Ordenar
    print(numeros);

    // Mapa

    print('\nMap:');
    Map<String, String> puesto = {'Mateo': 'Gerente', 'Marta': 'Desarrolladora de Software', 'Dani': 'Ventas'};
    print(puesto);

    puesto['Juan'] = 'Becario'; // Inserción
    print(puesto);

    puesto.remove('Dani'); // Borrado
    print(puesto);

    puesto['Juan'] = 'Ventas'; // Actualización
    print(puesto);

    Map<String, String> puestoOrdenado = SplayTreeMap<String, String>.from(puesto, (a, b) => a.compareTo(b)); // Ordenar
    print(puestoOrdenado);

    // Conjunto

    print('\nSet: ');
    Set<String> menu = {'Hamburgesa con queso', 'Patatas fritas', 'Resfresco a escoger'};
    print(menu);

    menu.add('Nuggets de pollo'); // Inserción
    print(menu);

    menu.remove('Patatas fritas'); // Borrado
    print(menu);

    /// Debido a que Set almacena datos únicos y no tiene un índice, la actualización directa no tiene una función.
    /// Lo mismo sucede con ordenarlo, para poder hacerlo, se ha de convertir en List y luego utilizar sort().
    
    List<String> menuActualizado = menu.toList(); // Actualización
    menuActualizado[1] = 'Pepsi';
    print(menuActualizado);


    List<String> menuOrdenado = menu.toList(); // Ordenar
    menuOrdenado.sort();
    print(menuOrdenado);
 

/*
 * DIFICULTAD EXTRA (opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
 *   los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no numéricos y con más de 11 dígitos.
 *   (o el número de dígitos que quieras)
 * - También se debe proponer una operación de finalización del programa.
 */

    Map<String, int> agenda = {};
    bool exit = false;

   while (!exit) {
      print('Bienvenido a la agenda de contactos. \n');
      print('Opciones: \n 1. Agregar \n 2. Buscar \n 3. Actualizar \n 4. Eliminar \n 5. Mostrar \n 6. Salir\n');
      print('Indica el número de la opción a realizar: ');
   
      String? opcion = stdin.readLineSync();

      switch (opcion) {
         case '1':

            print('Introduce el nombre: ');
            String? name = stdin.readLineSync();

            print('Introduce el número: ');
            String? number = stdin.readLineSync();

            if (name != null && number != null) {

               try {

                  int phone = int.parse(number);

                  if('$phone'.length <= 11) {
                     agregar(agenda, name, phone);
                  } else {
                     print('El número debe ser menor de 11 dígitos.');
                  }

               } catch (e) {
                  print('Introduce un número válido.');
               }
            } else {
               print('No se introdujo ningún número.');
            }
            break;
         
         case '2':
            
            print('Introduce el nombre del usuario que busca: ');
            String? name = stdin.readLineSync();

            if (name != null) {
               buscar(agenda, name);
            }
            break;

         case '3':
            
            print('Introduce el nombre: ');
            String? name = stdin.readLineSync();
            
            print('Introduce el número a actualizar: ');
            String? number = stdin.readLineSync();

            if (number != null && name != null) {
               try {
                  int phone = int.parse(number);
                  actualizar(agenda, name, phone);
               } catch (e) {
                  print('Introduce un número válido.');
               }
            }
            break;
         
         case '4':
            
            print('Introduce el nombre del contacto a eliminar: ');
            String? name = stdin.readLineSync();

            if (name != null) {
               eliminar(agenda, name);
            }
            break;
         
         case '5':
            
            mostrar(agenda);
            break;

         case '6':
            
            exit = true;
            break;
         
         default:
            print('Opción no válida.');
      }
   }
}

// Agregar

void agregar(Map<String, int> agenda, String name, int phone) {
    agenda[name] = phone;
    print('Contacto agregado.');
}

// Buscar

void buscar(Map<String, int> agenda, String name) {

   if(agenda.containsKey(name)) {
      'El número de $name es: ${agenda[name]}';
   } else {
      print('El contacto no existe');
   }
}

// Actualizar

void actualizar(Map<String, int> agenda, String name, int phone) {

   if(agenda.containsKey(name)) {   
      agenda[name] = phone;
      print('Contacto atualizado.');
   } else {
      print('El contacto no existe.');
   }
}
// Eliminar

void eliminar(Map<String, int> agenda, String name) {
   
   if (agenda.containsKey(name)) {
      agenda.remove(name);   
   } else {
      print('El contacto no existe.');
   }
}
// Mostrar

void mostrar(Map<String, int> agenda) {

   if (agenda.isEmpty) {
      print('La agenda está vacía.');
   } else {
      print(agenda);
   }
}